### PR TITLE
Update ComWrappers impl with best practices.

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
## Summary

Update the `ComWrappers` sample following the recommended guidance from the interop team.

### CODE REVIEW

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs [hunk 1]
```diff
@@ -13,31 +13,24 @@ internal unsafe class ComWrappersImpl : ComWrappers
...
+
+        // See IDispatch definition - https://docs.microsoft.com/windows/win32/api/oaidl/nn-oaidl-idispatch
+        var vtblRaw = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IDispatch), IntPtr.Size * 7);
+        vtblRaw[0] = fpQueryInterface;
+        vtblRaw[1] = fpAddRef;
+        vtblRaw[2] = fpRelease;
+        vtblRaw[3] = (IntPtr)(delegate* unmanaged<IntPtr, int*, int>)&IDispatchVtbl.GetTypeInfoCount;
+        vtblRaw[4] = (IntPtr)(delegate* unmanaged<IntPtr, int, int, IntPtr*, int>)&IDispatchVtbl.GetTypeInfo;
+        vtblRaw[5] = (IntPtr)(delegate* unmanaged<IntPtr, Guid*, IntPtr*, int, int, int*, int>)&IDispatchVtbl.GetIDsOfNames;
+        vtblRaw[6] = (IntPtr)(delegate* unmanaged<IntPtr, int, Guid*, int, ComTypes.INVOKEKIND, ComTypes.DISPPARAMS*, IntPtr, IntPtr, IntPtr, int>)&IDispatchVtbl.Invoke;
```
Comment (jkoritzinsky) [inline core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:29]: We could still use a typed vtable type with the function pointer types for the fields. That would make this slightly more readable.

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs [hunk 2]
```diff
@@ -13,31 +13,24 @@ internal unsafe class ComWrappersImpl : ComWrappers
...
-            GetIDsOfNames = Marshal.GetFunctionPointerForDelegate(IDispatchVtbl.pGetIDsOfNames),
-            Invoke = Marshal.GetFunctionPointerForDelegate(IDispatchVtbl.pInvoke)
-        };
-        var vtblRaw = RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IDispatchVtbl), sizeof(IDispatchVtbl));
-        Marshal.StructureToPtr(vtbl, vtblRaw, false);
-
-        wrapperEntry = (ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(IDispatchVtbl), sizeof(ComInterfaceEntry));
-        wrapperEntry->IID = typeof(IDispatch).GUID;
-        wrapperEntry->Vtable = vtblRaw;
+        GetIUnknownImpl(out IntPtr fpQueryInterface, out IntPtr fpAddRef, out IntPtr fpRelease);
```
Comment (kant2002) [inline core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:19]: I was suggested to use `GetIUnknownImpl(out vtblRaw[0], out vtblRaw[1], out vtblRaw[2]);` to reduce boilerplate previously in related discussions. Is this advice changed or here is inappropriate to show that to others?

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:29]: I don't think that actually adds much value. The type isn't needed in any way other development so if we are intending this sample to be the canonical way of how to implement this I would prefer to just implement it how a source generator should and avoid all superfluous metadata bloat.

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:19]: I can change it to that.

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs (hunk 1 above)
Comment (jkoritzinsky) [reply core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:29]: Yeah the main value of it is removing all of the casts here, but it's understandable if we don't want to go that route.

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:19]: Actually after looking at that I think I am going to keep it as is. The main reason is if someone adds another interface to support calling `GetIUnknownImpl` isn't needed and so will need to be teased out to reuse the pointers.

File: core/interop/comwrappers/IDispatch/ComWrappersImpl.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply core/interop/comwrappers/IDispatch/ComWrappersImpl.cs:29]: Agree the casting is annoying &ndash; I really dislike writing these manually.

**Review by jkotas — APPROVED** [review]

**Review by BillWagner — APPROVED** [review]
Thanks @AaronRobinsonMSFT 

I'll :shipit: now.

**Pull request merged**